### PR TITLE
[Torchvision API] crop

### PR DESCRIPTION
**Pull request by mdabek-nvidia**
## Category:

**New feature**

## Description:

Torchvision API operators RandomCrop and functional crop.
Unit tests checking conformance with Torchvision.
Known limitations: tensor and PIL.Image based paddings are not supported

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [X] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-4309

### CODE REVIEW

**Comment by mdabek-nvidia** [comment]
@greptileai please review

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            if key in fill:
+                return fill[key]
+        if "others" in fill:
+            return fill["others"]
+        raise ValueError(f"fill dictionary does not contain a value for {input_type}")
+
+    @staticmethod
+    def _randint(max_value):
+        range_start = fn.cast(max_value * 0, dtype=dali.types.FLOAT)
+        range_end = fn.cast(max_value + 1, dtype=dali.types.FLOAT)
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py:158]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+                fn.stack(
+                    fn.cast(-pad_left, dtype=dali.types.INT64),
+                    fn.cast(-pad_top, dtype=dali.types.INT64),
+                ),
+                fn.stack(in_w + pad_left + pad_right, in_h + pad_top + pad_bottom),
+                out_of_bounds_policy=PADDING_CLASS[self.padding_mode].border_type,
+                fill_values=self.fill,
+                device=self.device,
+                axis_names="WH",
+            )
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import numbers
+from types import NoneType
+from typing import Literal, Sequence, Union
+
+from PIL import Image
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+import nvidia.dali as dali
+import nvidia.dali.fn as fn
+import numpy as np
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+import torch
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: Fixed

**Comment by mdabek-nvidia** [comment]
@greptileai re-review please

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            padded_h = in_h + pad_top + pad_bottom
+            padded_w = in_w + pad_left + pad_right
+
+            if self.pad_if_needed:
+                pad_h = dali.math.max(crop_h - padded_h, 0)
+                pad_w = dali.math.max(crop_w - padded_w, 0)
+                pad_top = pad_top + pad_h
+                pad_bottom = pad_bottom + pad_h
+                pad_left = pad_left + pad_w
+                pad_right = pad_right + pad_w
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: Torchvision RandomCrop code is as follows:
```
# pad the width if needed
if self.pad_if_needed and width < self.size[1]:
    padding = [self.size[1] - width, 0]
    img = F.pad(img, padding, self.fill, self.padding_mode)
```
In the `F.pad`, `padding` parameter is extended to 4 elements as the documentation says:
`If sequence of length 2 is provided this is the padding on left/right and top/bottom respectively.`
So Torchvision behaves the same as this implementation 

I've also added a test to verify this: `test_random_crop_pad_if_needed_matches_torchvision_random_offsets` and it passes.

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+    @classmethod
+    def verify(cls, *, fill, **_) -> None:
+        if isinstance(fill, dict):
+            for key, value in fill.items():
+                if not isinstance(key, (type, str)):
+                    raise TypeError(f"fill dictionary keys must be types or strings, got {key!r}")
+                cls._verify_fill_value(value)
+        else:
+            cls._verify_fill_value(fill)
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: **[Correctness/UX]** `_ValidateFill.verify` accepts `dict` fill, and `adjust_fill` (line 155) handles it recursively — but `_kernel` (line 251) passes `self.fill` directly to `fn.slice(fill_values=...)`, which doesn't accept a dict. So a user who passes a dict (which torchvision *does* support) gets through validation cleanly and then hits a confusing late error from the DALI graph.

The commented-out tests at the bottom of `test_tv_randomcrop.py` confirm this is intentionally unsupported (`# TODO: Fill using dictionary pattern is currently not supported`).

Suggestion: either reject dict in `_ValidateFill.verify` with a clear `"fill via dict is not yet supported"` message, or wire the dict path through `_kernel`. Accepting at validation but failing later is the worst of both worlds. Also tied to greptile's earlier `_get_fill` dead-code finding — that cleanup removed the consumer but left the producers.

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        or a sequence of length 1 is provided this is used to pad all borders. If sequence of
+        length 2 is provided this is the padding on left/right and top/bottom respectively. If
+        a sequence of length 4 is provided this is the padding for the left, top, right and
+        bottom borders respectively.
+    pad_if_needed : bool, optional, default = False
+        Pad the image if it is smaller than the desired size.
+    fill : number or tuple or dict, optional, default = 0
+        Pixel fill value used when the padding_mode is constant.
+    padding_mode : Literal["constant", "edge", "reflect", "symmetric"], optional,
+        Type of padding. Should be: constant, edge, reflect or symmetric.
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py:113]: **[Nit/Docs]** The other parameters in this docstring all list their defaults (`default = None`, `default = False`, etc.). `padding_mode` is missing it — the signature defaults to `"constant"`. Suggest:

```
padding_mode : Literal["constant", "edge", "reflect", "symmetric"], optional, default = "constant"
    Type of padding. Should be: constant, edge, reflect or symmetric.
```

Also the trailing comma on line 122 (`, optional,`) reads as if something was cut off.

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+            padding_mode=padding_mode,
+            fill=fill,
+        )
+
+        self.size = RandomCrop.adjust_size(size)
+        self.padding = RandomCrop.adjust_padding(padding)
+        self.pad_if_needed = pad_if_needed
+        self.fill = RandomCrop.adjust_fill(fill)
+        self.padding_mode = padding_mode
+        self.needs_padding = pad_if_needed or any(self.padding)
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py:185]: **[Question]** `needs_padding = pad_if_needed or any(self.padding)` — when `pad_if_needed=True` but the input is already large enough, `pad_h = pad_w = 0` and the `fn.slice` block (lines 229-240) becomes an effective no-op that still runs in the graph (start=0, size=in_w/in_h). Worth a comment confirming that's intentional / cheap enough, or short-circuiting at graph construction time when both `any(self.padding) == False` and we don't actually need extra padding? Not blocking — just want to flag the extra graph node for large-input/`pad_if_needed=True` callers.

File: dali/test/python/torchvision/test_tv_randomcrop.py [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+
+def test_random_crop_fill_dict_matches_torchvision_pil():
+    img = make_pil_image(mode="RGB", h=4, w=5)
+    fill = {Image.Image: (1, 2, 3)}
+    _assert_equal_to_torchvision(
+        img,
+        _build_dali_random_crop(size=(6, 7), padding=1, fill=fill),
+        transforms.RandomCrop(size=(6, 7), padding=1, fill=fill),
+    )
+"""
```
Comment (jantonguirao) [inline dali/test/python/torchvision/test_tv_randomcrop.py]: **[Nit]** The two commented-out tests in a docstring-style block (lines 260-279) are easy to overlook and won't get picked up by lint as dead code. Suggest converting to `@unittest.expectedFailure` or `@nose2.tools.params([...])` + `@skip(reason="dict fill not supported")`, with a TODO message. That way the intent stays visible and the tests start running automatically when support lands.

**Review by jantonguirao — APPROVED** [review]

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py (hunk 9 above)
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py:185]: Added checking if padding is needed

File: dali/test/python/torchvision/test_tv_randomcrop.py (hunk 10 above)
Comment (mdabek-nvidia) [reply dali/test/python/torchvision/test_tv_randomcrop.py]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py (hunk 8 above)
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py:113]: Fixed, added default

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py (hunk 7 above)
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: Fixed

**Comment by mdabek-nvidia** [comment]
@greptileai, re-review

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+        pad_left, pad_top, pad_right, pad_bottom = self.padding
+
+        padded_h = in_h + pad_top + pad_bottom
+        padded_w = in_w + pad_left + pad_right
+
+        if self.pad_if_needed:
+            pad_h = dali.math.max(crop_h - padded_h, 0)
+            pad_w = dali.math.max(crop_w - padded_w, 0)
+            pad_top = pad_top + pad_h
+            pad_bottom = pad_bottom + pad_h
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+            pad_left = pad_left + pad_w
+            pad_right = pad_right + pad_w
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+        TODO: Fill using dictionary pattern is currently not supported
+        if isinstance(fill, dict):
+            for key, value in fill.items():
+                if not isinstance(key, (type, str)):
+                    raise TypeError(f"fill dictionary keys must be types or strings, got {key!r}")
+                cls._verify_fill_value(value)
+        else:
+            cls._verify_fill_value(fill)
+        """
+        cls._verify_fill_value(fill)
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py]: Fixed

**Comment by mdabek-nvidia** [comment]
@greptileai re-review

File: dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+            pad_left = pad_left + pad_w
+            # Only pad_top / pad_left are read below; pad_bottom / pad_right are dropped.
+            padded_h = padded_h + 2 * pad_h
+            padded_w = padded_w + 2 * pad_w
+
+        max_top = fn.cast(padded_h, dtype=dali.types.INT32) - crop_h
+        max_left = fn.cast(padded_w, dtype=dali.types.INT32) - crop_w
+
+        top = RandomCrop._randint(max_top)
+        left = RandomCrop._randint(max_left)
```
[the remark this answers is not in the document]
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/randomcrop.py:211]: Fixed in _randint, which in case of max_value < 0 will return (0,1) range.

**Comment by mdabek-nvidia** [comment]
@greptileai re-review

**Comment by mdabek-nvidia** [comment]
!build

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        return [-2, -1]
+    if inpt.layout[-2:] == "HW":
+        return [-2, -1]
+    raise ValueError(f"Unsupported layout: {inpt.layout!r}. Expected one of HWC, CHW, HW.")
+
+
+def _verify_crop_coordinate(value, name: str) -> None:
+    if not isinstance(value, int):
+        raise TypeError(f"{name} must be int, got {type(value)}")
+
```
Comment (mzient) [inline dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py:62]: Torchvision isn't that strict - despite type hints it does work with "ducktyped" ints.
```Python
>>> # This works fine...
>>> cropped = tvfn.crop(img, np.int32(4), np.int32(5), np.int32(100), np.int32(100))
>>> isinstance(np.int32(4), int)
False  # despite the type not being an `int`
```
In fact, it works just fine with floats and _rounds_ the _coordinates_, e.g.:
```Python
>>> tvfn.crop(img, 4.3, 5.2, 99.2, 100.2).size
(100, 100)
>>> tvfn.crop(img, 4.3, 5.2, 99.2, 100.3).size
(101, 100)
```
It computes the bounding box LTRB in whatever type was given, then rounds.
In the first case we get:
4.3, 5.2, 103.5, 105.4, which gets rounded to 4, 5, 104, 105
and in the second we get
4.3, 5.2, 103.5, 105.5, which gets rounded to 4, 5, 104, 106

Comment (mzient) [reply dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py:62]: Update: this only happens with PIL.Image inputs, but still - it does happen.

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        pad_if_needed=False,
+        padding_mode="constant",
+        fill=0,
+    )
+
+    return ndd.slice(
+        inpt,
+        [float(top), float(left)],
+        [float(height), float(width)],
+        axes=_get_crop_axes(inpt),
```
Comment (mzient) [inline dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py]: This looks weird - since we already assume that _get_crop_axes is going to find `HW` in the layout, why not just go with the default layout handling, like we do in the case of `fn` API?
We could verify that the layout contains `HW` substring and that's it, slice will handle the rest.
```suggestion
        [float(left), float(top)],
        [float(width), float(height)],
```

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py (hunk 15 above)
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py:62]: Fixed

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py (hunk 16 above)
Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py]: Fixed

**Comment by mdabek-nvidia** [comment]
!build

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    left: int | float,
+    height: int | float,
+    width: int | float,
+    device: DeviceLike = "cpu",
+) -> ndd.Tensor | ndd.Batch:
+    """
+    Please refer to the ``RandomCrop`` operator for more details.
+    """
+    top, left, height, width = _validate_crop_params(inpt, top, left, height, width)
+    RandomCrop.verify_args(
```
Comment (mzient) [inline dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py:77]: Nit:
```suggestion
    RandomCrop.validate_args(
```

Comment (mdabek-nvidia) [reply dali/python/nvidia/dali/experimental/torchvision/v2/functional/crop.py:77]: It requires refactoring the Operator class, which leads to changing almost all of the objective operators. I can do it in a separate PR.

**Review by mzient — APPROVED** [review]
